### DISCUSSION TITLE
Lifetime choice for Scanner

### DISCUSSION THREAD

**Discussion by potto216** [DireLines/lox-rs#2] [Q&A] [answered]
Is this accurate? To make sure I understand the lifetime choice for Scanner I am summarizing our discussions from 8/10 which focused on the [Scanner code](https://github.com/DireLines/lox-rs/blob/a011a5d1e1c1be02a1d3c4a70394b7900192061f/treewalk/src/main.rs#L50).

So for example the `Scanner` code:
``` Rust
impl<'a> Scanner<'a> {
    pub fn new(source: &'a str) -> Self {
        Scanner {
            source,
            next_unparsed: 0,
            line: 0,
        }
    }
}
```
The lifetime `a` is necessary because struct data is borrowed and `source` should not outlive the the data `str` that it is referencing.  If `source` was a clone of `str` then the lifetime `a` would not be needed.

`impl<'a>` is needed to show the relationship in lifetimes between multiple data structures. For example the `Iterator` trait is implemented for the `Scanner` struct and returns a `Token` struct. Both the `Scanner` struct and `Token` struct reference data with lifetimes. `impl<'a>` lets the Rust compiler know that they are the same lifetime

**Comment by DireLines** [comment, marked as answer]
Well, of course you need to define `'a` somewhere scoped to the impl block so you know what it means inside the parameters for `new`, but I was wondering why you would need `impl<'a> Scanner<'a>` and not simply `impl Scanner<'a>`. After all, the compiler should know what `'a` means since it has access to the definition of `Scanner` which has a generic parameter.

I asked some people on a rust discord server about this, and they gave some very illustrative examples of why the generic types need to be specified in the `impl` in some cases:

You can write both generic and specialized implementations for a type:
```rust
struct Foo<T>(T);
// generic implementation for any type T
impl<T> Foo<T> {}
//but I want different behavior for the specific type i32
impl Foo<i32> {}
```
If the name `T` already refers to a type in scope, you need to be able to distinguish between a specialized impl for `T` and a generic implementation:
```rust
struct Foo<T>(T);

struct T;

// this is an impl that exists specifically for the concrete type named T
impl Foo<T> {}

// this is an impl that exists for any type
impl<T> Foo<T> {}
```

Also, the generic parameter for the impl can show up in other places besides the generic parameters on the struct, for example
```rust
impl<T> Trait for T {} //blanket implementation of the trait Trait
impl<T> Trait<T> for Other {} //T is a parameter of the trait, not the struct
```

I asked GPT4 first and it gave some terrible examples where `impl<T>` was still redundant and quickly gave up trying to convince me after I pointed out why

**Reply by DireLines** [reply]
I feel as though Rust is choosing explicitness / genericness / flexibility over having sane defaults here. In these atypical cases it makes sense to have the types specified after `impl`, but in the typical case you have some struct `MyStruct<A, B, C>` and now you want to implement some methods using the *exact same* generic types with the *exact same* trait bounds as the ones you specified in the definition of the struct. In that case, saying `impl<A, B, C> MyStruct<A, B, C>` is totally redundant. Rust is making the atypical case expressible by making the typical case worse.

**Reply by potto216** [reply]
@DireLines thanks for the explanation and examples. Did the people on the Rust Discord server give any examples for the benefit of specifying the lifetime in impl<>?

**Reply by DireLines** [reply]
Yeah, those are the examples. The lifetime is a generic parameter just like `T` in `impl<T> MyStruct<T>`. It's not so much a benefit as it is a format forced by the compiler to have uniform syntax for all these different cases.
